dvc/external_repo.py

```diff
@@ -1,35 +1,43 @@
 import logging
 import os
 import tempfile
-from contextlib import contextmanager
+from contextlib import contextmanager, suppress
 from distutils.dir_util import copy_tree
 import threading
 
-from funcy import retry, suppress, wrap_with, cached_property
+from funcy import retry, wrap_with, cached_property
 
 from dvc.path_info import PathInfo
 from dvc.compat import fspath
 from dvc.repo import Repo
+from dvc.repo.tree import RepoTree
 from dvc.config import NoRemoteError, NotDvcRepoError
 from dvc.exceptions import NoRemoteInExternalRepoError, CheckoutError
 from dvc.exceptions import OutputNotFoundError, NoOutputInExternalRepoError
-from dvc.exceptions import FileMissingError, PathMissingError
-from dvc.utils.fs import remove, fs_copy, move
+from dvc.exceptions import FileMissingError, PathMissingError, DownloadError
+from dvc.utils.fs import makedirs, move, remove
 from dvc.utils import tmp_fname
 from dvc.scm.git import Git
+from dvc.scm.tree import WorkingTree
 
 
 logger = logging.getLogger(__name__)
 
 
 @contextmanager
-def external_repo(url, rev=None, for_write=False):
+def external_repo(
+    url, rev=None, for_write=False, cache_dir=None, cache_types=None
```

Comment:
I purposefully removed cache args from here. This way all the intermediate code needs to know about them, which makes things more complex overall.

Why do you change it back? Can we set it later?

Reply:
It seemed to me that rather than having
```
    with external_repo(...) as repo:
            if hasattr(repo, "cache"):
                repo.cache.local.cache_dir = self.repo.cache.local.cache_dir
```
everywhere we use erepos, we should just handle that in `external_repo()`, since it depends on whether or not we return an `ExternalRepo` or `GitExternalRepo`.

With the previous behavior, for `ExternalRepo` we always call `_set_cache_dir()` which sets `cache_dir` to a cached temp directory. And then later we would sometimes change `cache_dir` to some another existing cache directory.

It seemed to me that it makes more sense to just have `_set_cache_dir()` set it to the directory we want right away, rather than setting it to a temp dir that we don't actually need, and then changing it again later.

Reply:
After looking into this some more, I get why it was structured this way now, and my changes have been reverted